### PR TITLE
env_process: Separate image check and remove in postprocess

### PR DESCRIPTION
**Pull request by hereischen**
It will check all the images at one time in image check
procedure, then delete images need to be remove.
This will solve problems occurred when checking and removing
chained images one by one.

Signed-off-by: Haotong Chen <hachen@redhat.com>

id: 1560377

### CODE REVIEW

File: virttest/env_process.py [hunk 1]
```diff
@@ -341,9 +341,26 @@ def postprocess_image(test, params, image_name, vm_process_status=None):
                 logging.warn(e.message)
             else:
                 raise e
-    if (not restored and
-            params.get("restore_image_after_testing", "no") == "yes"):
-        image.backup_image(params, base_dir, "restore", True)
+        if (not restored and
+                params.get("restore_image_after_testing", "no") == "yes"):
+            image.backup_image(params, base_dir, "restore", True)
```
Comment (pingl2016) [inline virttest/env_process.py]: Code block for restoring images should be moved to postprocess_image()

File: virttest/env_process.py [hunk 2]
```diff
@@ -563,12 +580,42 @@ def _call_image_func():
         else:
             process_images(image_func, test, params)
 
+    def _call_check_image_func():
```
Comment (pingl2016) [inline virttest/env_process.py]: Try to reduce the redundant code between _call_image_func() and _call_check_image_func()

Comment (hereischen) [reply virttest/env_process.py]: Thanks.
This patch serves as a work around of the problem.
It needs to be refactored later on.

File: virttest/env_process.py [hunk 3]
```diff
@@ -319,32 +319,60 @@ def postprocess_image(test, params, image_name, vm_process_status=None):
...
-            if params.get("remove_image_on_check_error", "no") == "yes":
-                cl_images = params.get("master_images_clone", "")
-                if image_name in cl_images.split():
-                    image.remove()
             if (params.get("skip_cluster_leak_warn") == "yes" and
                     "Leaked clusters" in e.message):
                 logging.warn(e.message)
             else:
-                raise e
+                params["img_check_failed"] = "yes"
```
[the remark this answers is not in the document]
Comment (hereischen) [reply virttest/env_process.py]: It serves as a work around. Will try to move it as an attribute of `img` object. Thanks

File: virttest/env_process.py (hunk 1 above)
Comment (hereischen) [reply virttest/env_process.py]: updated, thanks

File: virttest/env_process.py [hunk 4]
```diff
@@ -319,32 +318,60 @@ def postprocess_image(test, params, image_name, vm_process_status=None):
...
             params.get("restore_image_after_testing", "no") == "yes"):
         image.backup_image(params, base_dir, "restore", True)
-    if params.get("remove_image", "yes") == "yes":
+
+    if params.get("img_check_failed") == "yes":
+        if params.get("remove_image_on_check_error", "no") == "yes":
+            cl_images = params.get("master_images_clone", "")
+            if image_name in cl_images.split():
+                image.remove()
+                restored = True
```
Comment (luckyh) [inline virttest/env_process.py]: Should be `removed = True`

File: virttest/env_process.py [hunk 5]
```diff
@@ -319,32 +318,60 @@ def postprocess_image(test, params, image_name, vm_process_status=None):
...
+        if params.get("restore_image_on_check_error", "no") == "yes":
+            image.backup_image(params, base_dir, "restore", True)
+            restored = True
+    else:
+        # Allow test to overwrite any pre-testing  automatic backup
+        # with a new backup. i.e. assume pre-existing image/backup
+        # would not be usable after this test succeeds. The best
+        # example for this is when 'unattended_install' is run.
+        if params.get("backup_image", "no") == "yes":
+            image.backup_image(params, base_dir, "backup", True)
```
Comment (luckyh) [inline virttest/env_process.py]: Should skip the restore operation when vm is running.

Comment (hereischen) [reply virttest/env_process.py]: all updated, thanks

**Comment by hereischen** [comment]
(1/2) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.qemu_img_negative.rebase.nonexist_old_backing_file: PASS (5.35 s)
 (2/2) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.qemu_img_negative.rebase.nonexist_new_backing_file: PASS (5.45 s)

File: virttest/env_process.py [hunk 6]
```diff
@@ -563,13 +590,47 @@ def _call_image_func():
...
+                    images = params.objects("images")
+                    _process_images_serial(
+                        check_image, test, images, vm_params,
+                        vm_process_status=vm_process_status)
+                finally:
+                    if unpause_vm:
+                        vm.resume()
+        else:
+            images = params.objects("images")
+            _process_images_serial(check_image, test, images, vm_params)
```
Comment (hereischen) [inline virttest/env_process.py]: should pass `params` into the function here.

**Comment by luckyh** [comment]
Let's merge this, thanks.

**Pull request merged**